---
title: "How to draw gene structure graph and mapping primer position/direction?"
output: html_document
---
################
##      ver0.4 
## by Kazunari Nozue
################
* ver0.2 is OK version.  
* needs to fix inconsistent strand color between geen structure and primers (050715)  
* idea1: use all TxDx object to draw and specify gene or primers by "which" arguments that should be GRange object (ver0.3)  
* idea2: combine all GR object and convert them into GRL. (050815-, v0.4), which works OK.
* clean up scripts (051115 and making version 2 script)  

# preparation
```{r eval=FALSE} 
# run this section once
# preparation
source("http://bioconductor.org/biocLite.R")
biocLite(ggbio)
# making Arabidopsis thaliana TxDb package
library("org.At.tair.db")
class(org.At.tair.db)
# columns(org.At.tair.db) # http://www.bioconductor.org/help/workflows/annotation/annotation/
# keytypes(org.At.tair.db)
# ids<-head(keys(org.At.tair.db, keytype="ENTREZID"))
# select(org.At.tair.db, keys=ids, columns="SYMBOL", keytype="ENTREZID")

 TxDb.At<-makeTxDbFromGFF("/Volumes/Data8/NGS_related/Arabidopsis_analysis/reference/TAIR10_GFF3_genes.gff",format="gff3",organism="Arabidopsis thaliana")
 makeTxDbPackage(txdb=TxDb.At,
                 version="1.0.0",
                maintainer="Kazunari Nozue <knozue@ucavis.edu>",
                author="Kazunari Nozue",
                destDir="/Volumes/Data6/bioconductor_R")
system("tar -zcvf TxDb.Athaliana.tar.gz TxDb.Athaliana") #in /Volumes/Data6/bioconductor_R
install.packages("/Volumes/Data6/bioconductor_R/TxDb.Athaliana.tar.gz", repos = NULL, type = "source")
```

# primer design and get position information in a genome
Use http://www.ncbi.nlm.nih.gov/tools/primer-blast/  
* Enter AGI name in NCBI > gene  
* click "graphic" at the top  
* Find mRNA name on the bottom (eg. )  
* Click the mRNA name.  
* click "graphic"  
* click "tools" in the tool bar and go to primer-NCBI  
* enter primer sequences and run primer-NCBI  
* You should have PCR region within mRNA  
![image](/Volumes/Data6/data_JM4/qRT-PCR_2015/primer_position/Screen%20Shot%202015-05-08%20at%201.00.24%20PM.png)


```{r eval=TRUE}
setwd("/Volumes/Data6/data_JM4/qRT-PCR_2015/primer_position")
library("TxDb.Athaliana", lib.loc="/Library/Frameworks/R.framework/Versions/3.2/Resources/library")
library(ggbio)
library(biovizBase)
# cf http://www.tengfei.name/ggbio/docs/man/autoplot-method.html

# read primer info file
primer.info<-read.csv("/Volumes/Data6/data_JM4/paper manuscript_data_JM4/timecourseSASRNAseq/primer.info2015timecourseSAS.csv")

# ver0.4 function
# combine gene structure and primer as one GRL object
primer.gene.plot.v0.4<-function(primers,AGI) {
  # making primer GRange object 
  primers$primername<-as.character(primers$primername)
  # add primer info for AT1G01060 (fake primers)
  primer1 <- GRanges(paste("Chr",primers[1,"Chr"],sep=""), IRanges(primers[1,"start"], primers[1,"end"]),strand=as.character(primers[1,"strand"]),tx_id="",tx_name=primers[1,"primername"],gene_id=primers[1,"AGI"],model="utr")
  primer2 <- GRanges(paste("Chr",primers[2,"Chr"],sep=""), IRanges(primers[2,"start"], primers[2,"end"]),strand=as.character(primers[2,"strand"]),tx_id="",tx_name=primers[2,"primername"],gene_id=primers[2,"AGI"],model="utr") 
  # if primer is splitted
  if(!primers[1,"start2"]=="N/A")  {
        print(primers[1,])
    primer1b<-GRanges(paste("Chr",primers[1,"Chr"],sep=""), IRanges(as.numeric(as.character(primers[1,"start2"])), as.numeric(as.character(primers[1,"end2"]))),strand=as.character(primers[1,"strand"]),tx_id="",tx_name=primers[1,"primername"],gene_id=primers[1,"AGI"],model="utr")
    primer1<-c(primer1,primer1b)
    }
  if(!primers[2,"start2"]=="N/A")  {
    print(primers[2,])
      primer2b<-GRanges(paste("Chr",primers[2,"Chr"],sep=""), IRanges(as.numeric(as.character(primers[2,"start2"])), as.numeric(as.character(primers[2,"end2"]))),strand=as.character(primers[2,"strand"]),tx_id="",tx_name=primers[2,"primername"],gene_id=primers[2,"AGI"],model="utr")
    primer2<-c(primer2,primer2b)
    }
  primers.temp<-c(primer1,primer2) #GRange object
  #print(strand(primers.temp))
  print(primers.temp) 
#
 transcripts.Athaliana.selected.temp<-transcripts(TxDb.Athaliana,filter=list(gene_id=AGI))
  gr.Athaliana.temp <- crunch(TxDb.Athaliana,which=transcripts.Athaliana.selected.temp)
  ## change column to 'model'
  colnames(values(gr.Athaliana.temp))[4] <- "model"
  # into GRL
  gr.Athaliana.primer.temp<-c(gr.Athaliana.temp,primers.temp)
  # into GRL
  grl.Athaliana.primer.temp <- split(gr.Athaliana.primer.temp, gr.Athaliana.primer.temp$tx_name)
  # print 
  print(grl.Athaliana.primer.temp)
  # plot
  cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
    strand.color <- c("magenta", "black", "green")


  q<-autoplot(grl.Athaliana.primer.temp,aes(type=model,fill=strand))+ scale_fill_manual(values= strand.color) 
  q<-q + theme(axis.text.y=element_blank(),panel.grid.major = element_blank(), panel.grid.minor =           element_blank(), 
               panel.background = element_blank(), axis.line.y = element_line(colour = "white"),
               axis.ticks.y=element_line(color="white"))#,
               #legend.position = "none")
  return(q)
}
```
# plot all genes in primer.info table
```{r eval=TRUE}
setwd("/Volumes/Data6/data_JM4/qRT-PCR_2015/primer_position")
AGI.levels<-unique(primer.info[37:38,"AGI"])
for(x in AGI.levels) {
  r<-primer.gene.plot.v0.4(primers=primer.info[primer.info$AGI==x,],AGI=x)
  splicing.varians<-transcripts.Athaliana.selected.temp<-transcripts(TxDb.Athaliana,filter=list(gene_id=x))
print(paste(x,"_primer_position.v0.4.png",sep=""))
  ggsave(file=paste(x,"_primer_position.v0.4.png",sep=""),height=length(splicing.varians$tx_name)+1,width=8,dpi=300,units="in")
}

```

```{r eval=TRUE}
sessionInfo()
```

